### DISCUSSION TITLE
Getting consumption power figure?

### DISCUSSION THREAD

**Discussion by cajsoft** [ZuinigeRijder/SolisCloud2PVOutput#24] [General]
Hi, I'd like to get the self consumption figure from the API but cant seem to find the correct one.  I have solis hybrid inverter.

I tried [oneSelf] but it doesnt match the figure in the image below.

![image](https://github.com/ZuinigeRijder/SolisCloud2PVOutput/assets/33936696/6a32c24d-c725-4f1b-adbd-71bb016a7bb0)

thanks
Craig.

**Comment by ZuinigeRijder** [comment]
Probably you have seen this post, about the Power Used:

https://github.com/ZuinigeRijder/SolisCloud2PVOutput/issues/15#issuecomment-1499063092

But I assume you also want to know the battery in/out? So you can compute the self consumption?
From the [SolisCloud API ](https://github.com/ZuinigeRijder/SolisCloud2PVOutput/blob/main/SolisCloud%20Monitoring%20API1.2.pdf)I think you should get from /v1/api/inveterDetail: batteryPower

batteryPower= round(inverter_detail['batteryPower'] * 1000)

And there is also bypassLoadPower.
With the other information, including familyLoad, you can compute the self consumption figure?

Just play around, to see if you can come up with the computation to get the above figures.
And if yes, please post your code, so others can also use it.

**Reply by cajsoft** [reply]
Hi, thanks for the info, but I dont think that's the correct attributes.. the live consumption figure is generally taken via the CT clamp on the eastron meter. this meter sends consumption data to solis hybrid inverter.   This must be able to be pulled from the API as the figure shows on the app and website.  This consumption figure has nothing to do with the battery.   If battery was removed the figure would still be present as it represents the house load.  Will keep digging...

Craig.

**Comment by ZuinigeRijder** [comment]
Also not the bypassLoadPower?

**Reply by cajsoft** [reply]
both figures return 0

**Comment by ZuinigeRijder** [comment]
Maybe make a screenshot as before and at the same time run the adapted script to see the values:

Add after line 248, just before "if SEND_TO_PVOUTPUT": 
[code]
            pSum = round(inverter_detail['pSum'] * 1000)
            batteryPower = round(inverter_detail['batteryPower'] * 1000)
            bypassLoadPower = round(inverter_detail['bypassLoadPower'] * 1000)
            familyLoadPower = round(inverter_detail['familyLoadPower'] * 1000)
            print(f"hi_res_watthour_today={hi_res_watthour_today}, watt={watt}, pSum={pSum}, familyLoadPower={familyLoadPower}, bypassLoadPower={bypassLoadPower}, batteryPower={batteryPower}")
[/code]

Maybe we can figure out the pieces of the puzzle.

**Reply by cajsoft** [reply]
ok will try this shortly and post back :)

**Reply by cajsoft** [reply]
![image](https://github.com/ZuinigeRijder/SolisCloud2PVOutput/assets/33936696/a00877d2-becb-4bac-97e4-6cd3315d9695)

seems pSum is not valid.

**Comment by ZuinigeRijder** [comment]
Ah, ok. Then let's try to just dump all inverter_detail values.
Instead of the code to add, just uncomment line 201 and 202 by removing the hash # character.

[code]
        json_formatted_str = json.dumps(inverter_detail, indent=2)
        print(json_formatted_str)
[/code]

And also make a screenshot at the same time you dump the inverter_details.

**Reply by cajsoft** [reply]
![image](https://github.com/ZuinigeRijder/SolisCloud2PVOutput/assets/33936696/4be0e9ee-0d74-4b56-9ed6-2a9c0cebabe6)

[code]
"fullHourStr": "h",                                                                            
  "currentState": "3",                                                                                                                                                  
  "warningInfoData": 0,                                                                                                                                                 
  "timeZone": 0.0,                                                                                                                                                      
  "timeZoneStr": "UTC+00:00",                                                                                                                                           
  "daylight": 3600000,                                                                                                                                                  
  "daylightSwitch": 1,                                                                                                                                                  
  "model": "3104",                                                                                                                                                      
  "productModel": "3104",                                                                                                                                               
  "ctrlCommand": 2,                                                                                                                                                     
  "nationalStandardstr": "G99",                                                                                                                                         
  "inverterTemperature": 44.7,                                                                                                                                          
  "inverterTemperatureUnit": "\u2103",                                                                                                                                  
  "temp": 105.0,                                                                                                                                                        
  "tempName": "Inverter Internal Operating Ambient Temperature",                                                                                                        
  "sno": "159044",                                                                                                                                                      
  "stationId": "1298491919448948139",                                                                                                                                   
  "version": "420032",                                                                                                                                                  
  "acOutputType": 0,                                                                                                                                                    
  "dcInputtype": 1,                                                                                                                                                     
  "rs485ComAddr": "101",                                                                                                                                                
  "dataTimestamp": "1684500488028",                                                                                                                                     
  "timeStr": "2023-05-19 13:48:08",                                                                                                                                     
  "reactivePower": 0.0,                                                                                                                                                 
  "apparentPower": 160.0,                                                                                                                                               
  "dcPac": 0.0,                                                                                                                                                         
  "uInitGnd": 0,                                                                                                                                                        
  "uInitGndStr": "V",                                                                                                                                                   
  "dcBus": 389.1,                                                                                                                                                       
  "dcBusStr": "V",                                                                                                                                                      
  "dcBusHalf": 0.0,                                                                                                                                                     
  "dcBusHalfStr": "V",                                                                                                                                                  
  "power": 5.0,                                                                                                                                                         
  "powerStr": "kWp",                                                                                                                                                    
  "powerPec": "1",                                                                                                                                                      
  "porwerPercent": 0.424,                                                                                                                                               
  "pac": 2.122,                                                                                                                                                         
  "pacStr": "kW",                                                                                                                                                       
  "pacPec": "1",                                                                                                                                                        
  "oneSelf": 331.9,                                                                                                                                                     
  "eToday": 14.4,                                                                                                                                                       
  "eTodayStr": "kWh",                                                                                                                                                   
  "eMonth": 472.0,                                                                                                                                                      
  "eMonthStr": "kWh",                                                                                                                                                   
  "eYear": 1.774,                                                                                                                                                       
  "eYearStr": "MWh",                                                                                                                                                    
  "eTotal": 1.774,                                                                                                                                                      
  "eTotalStr": "MWh",                                                                                                                                                   
  "uPv1": 271.3,                                                                                                                                                        
  "uPv1Str": "V",                                                                                                                                                       
  "iPv1": 3.9,                                                                                                                                                          
  "iPv1Str": "A",                                                                                                                                                       
  "uPv2": 280.2,                                                                                                                                                        
  "uPv2Str": "V",                                                                                                                                                       
  "iPv2": 3.8,                                                                                                                                                          
  "iPv2Str": "A",        
"uAc1": 247.7,                                                                                                                                                        
  "uAc1Str": "V",                                                                                                                                                       
  "iAc1": 2.1,                                                                                                                                                          
  "iAc1Str": "A",                                                                                                                                                       
  "uAc2": 0.0,                                                                                                                                                          
  "uAc2Str": "V",                                                                                                                                                       
  "iAc2": 0.0,                                                                                                                                                          
  "iAc2Str": "A",                                                                                                                                                       
  "uAc3": 0.0,                                                                                                                                                          
  "uAc3Str": "V",                                                                                                                                                       
  "iAc3": 0.0,                                                                                                                                                          
  "iAc3Str": "A",                                                                                                                                                       
  "powerFactor": 0.0,                                                                                                                                                   
  "batteryDischargeEnergy": 0,                                                                                                                                          
  "batteryDischargeEnergyStr": "kWh",                                                                                                                                   
  "batteryChargeEnergy": 0,                                                                                                                                             
  "batteryChargeEnergyStr": "kWh",                                                                                                                                      
  "homeLoadEnergy": 0,                                                                                                                                                  
  "homeLoadEnergyStr": "kWh",                                                                                                                                           
  "gridPurchasedEnergy": 0,                                                                                                                                             
  "gridPurchasedEnergyStr": "kWh",                                                                                                                                      
  "gridSellEnergy": 0,                                                                                                                                                  
  "gridSellEnergyStr": "kWh",                                                                                                                                           
  "fac": 50.08,                                                                                                                                                         
  "facStr": "Hz",                                                                                                                                                       
  "batteryPower": 1.79,                                                                                                                                                 
  "batteryPowerStr": "kW",                                                                                                                                              
  "batteryPowerPec": "1",                                                                                                                                               
  "batteryPowerZheng": 1790.1,                                                                                                                                          
  "batteryPowerFu": 0,                                                                                                                                                  
  "storageBatteryVoltage": 51.0,                                                                                                                                        
  "storageBatteryVoltageStr": "V",                                                                                                                                      
  "storageBatteryCurrent": 35.1,                                                                                                                                        
  "storageBatteryCurrentStr": "A",                                                                                                                                      
  "batteryCapacitySoc": 83.0,                                                                                                                                           
  "batteryHealthSoh": 100.0,                                                                                                                                            
  "batteryVoltage": 50.05,                                                                                                                                              
  "batteryVoltageStr": "V",                                                                                                                                             
  "bstteryCurrent": 35.1,                                                                                                                                               
  "bstteryCurrentStr": "A",                                                                                                                                             
  "batteryPowerBms": 1.757,                                                                                                                                             
  "batteryPowerBmsStr": "kW",                                                                                                                                           
  "internalBatteryI": 0.0,                                                                                                                                              
  "batteryChargingCurrent": 80.0,                                                                                                                                       
  "batteryChargingCurrentStr": "A",                                                                                                                                     
  "batteryDischargeLimiting": 100.0,                                                                                                                                    
  "batteryDischargeLimitingStr": "A",                                                                                                                                   
  "batteryFailureInformation01": "0",                                                                                                                                   
  "batteryFailureInformation02": "0",                                                                                                                                   
  "batteryTotalChargeEnergy": 621.0,                                                                                                                                    
  "batteryTotalChargeEnergyStr": "kWh",                                                                                                                                 
  "batteryTodayChargeEnergy": 3.9,                                                                                                                                      
  "batteryTodayChargeEnergyStr": "kWh",                                                                                                                                 
  "batteryMonthChargeEnergy": 0,                                                                                                                                        
  "batteryMonthChargeEnergyStr": "kWh",                                                                                                                                 
  "batteryYearChargeEnergy": 0,                                                                                                                                         
  "batteryYearChargeEnergyStr": "kWh",                                                                                                                                  
  "batteryYesterdayChargeEnergy": 3.5,                                                                                                                                  
  "batteryYesterdayChargeEnergyStr": "kWh",                                                                                                                             
  "batteryTotalDischargeEnergy": 682.0,                                                                                                                                 
  "batteryTotalDischargeEnergyStr": "kWh",                                                                                                                              
  "batteryTodayDischargeEnergy": 3.9,                                                                                                                                   
  "batteryTodayDischargeEnergyStr": "kWh",                                                                                                                              
  "batteryMonthDischargeEnergy": 0,                                                                                                                                     
  "batteryMonthDischargeEnergyStr": "kWh",                                                                                                                              
  "batteryYearDischargeEnergy": 0,                                                                                                                                      
  "batteryYearDischargeEnergyStr": "kWh",                                                                                                                               
  "batteryYesterdayDischargeEnergy": 3.9,                                                                                                                               
  "batteryYesterdayDischargeEnergyStr": "kWh",                                                                                                                          
  "gridPurchasedTotalEnergy": 356.81,                                                                                                                                   
  "gridPurchasedTotalEnergyStr": "kWh",                                                                                                                                 
  "gridPurchasedYearEnergy": 0,                                                                                                                                         
  "gridPurchasedYearEnergyStr": "kWh",                                                                                                                                  
  "gridPurchasedMonthEnergy": 0,                                                                                                                                        
  "gridPurchasedMonthEnergyStr": "kWh",                                                                                                                                 
  "gridPurchasedTodayEnergy": 0.3,                                                                                                                                      
  "gridPurchasedTodayEnergyStr": "kWh",                                                                                                                                 
  "gridPurchasedYesterdayEnergy": 0.3,                                                                                                                                  
  "gridPurchasedYesterdayEnergyStr": "kWh",                                                                                                                             
  "gridSellTotalEnergy": 567.16,                                                                                                                                        
  "gridSellTotalEnergyStr": "kWh",                                                                                                                                      
  "gridSellYearEnergy": 0,                                                                                                                                              
  "gridSellYearEnergyStr": "kWh",                                                                                                                                       
  "gridSellMonthEnergy": 0,                                                                                                                                             
  "gridSellMonthEnergyStr": "kWh",                                                                                                                                      
  "gridSellTodayEnergy": 2.6,                                                                                                                                           
  "gridSellTodayEnergyStr": "kWh",                                                                                                                                      
  "gridSellYesterdayEnergy": 12.1,                                                                                                                                      
  "gridSellYesterdayEnergyStr": "kWh",                                                                                                                                  
  "homeLoadTotalEnergy": 1.606,                                                                                                                                         
  "homeLoadTotalEnergyStr": "MWh",                                                                                                                                      
  "homeLoadTodayEnergy": 12.0,                                                                                                                                          
  "homeLoadTodayEnergyStr": "kWh",                                                                                                                                      
  "totalLoadPower": 0,                                                                                                                                                  
  "totalLoadPowerStr": "kW",                                                                                                                                            
  "homeLoadYesterdayEnergy": 7.7,                                                                                                                                       
  "homeLoadYesterdayEnergyStr": "kWh",                                                                                                                                  
  "familyLoadPower": 0.179,                                                                                                                                             
  "familyLoadPowerStr": "kW",                                                                                                                                           
  "familyLoadPercent": 0,                                                                                                                                               
  "homeGridYesterdayEnergy": 0.0,                                                                                                                                       
  "homeGridYesterdayEnergyStr": "kWh",                                                                                                                                  
  "homeGridTodayEnergy": 0.0,                                                                                                                                           
  "homeGridTodayEnergyStr": "kWh",                                                                                                                                      
  "homeGridMonthEnergy": 0.0,                                                                                                                                           
  "homeGridMonthEnergyStr": "kWh",                                                                                                                                      
  "homeGridYearEnergy": 0.0,                                                                                                                                            
  "homeGridYearEnergyStr": "kWh",                                                                                                                                       
  "homeGridTotalEnergy": 0.0,                                                                                                                                           
  "homeGridTotalEnergyStr": "kWh",                                                                                                                                      
  "bypassLoadPower": 0.0,                                                                                                                                               
  "bypassLoadPowerStr": "kW",                                                                                                                                           
  "backupYesterdayEnergy": 0.0,                                                                                                                                         
  "backupYesterdayEnergyStr": "kWh",                                                                                                                                    
  "backupTodayEnergy": 0.0,                                                                                                                                             
  "backupTodayEnergyStr": "kWh",                                                                                                                                        
  "backupMonthEnergy": 0.0,                                                                                                                                             
  "backupMonthEnergyStr": "kWh",                                                                                                                                        
  "backupYearEnergy": 0.0,                                                                                                                                              
  "backupYearEnergyStr": "kWh",                                                                                                                                         
  "backupTotalEnergy": 0.0,                                                                                                                                             
  "backupTotalEnergyStr": "kWh",                                                                                                                                        
  "bypassAcVoltage": 247.6,                                                                                                                                             
  "bypassAcVoltageB": 0.0,                                                                                                                                              
  "bypassAcVoltageC": 0.0,                                                                                                                                              
  "bypassAcCurrent": 2.1,                                                                                                                                               
  "bypassAcCurrentB": 0.0,                                                                                                                                              
  "bypassAcCurrentC": 0.0,                                                                                                                                              
  "pLimitSet": 110.0,                                                                                                                                                   
  "pFactorLimitSet": 10.0,                                                                                                                                              
  "pReactiveLimitSet": 0.0,                                                                                                                                             
  "batteryType": "4.0",                                                                                                                                                 
  "socDischargeSet": 15.0,                                                                                                                                              
  "socChargingSet": 10.0,                                                                                                                                               
  "pEpmSet": 0.0,                                                                                                                                                       
  "pEpmSetStr": "kW",                                                                                                                                                   
  "epmFailSafe": 0.0,                                                                                                                                                   
  "epmSafe": 0,                                                                                                                                                         
  "pEpm": 0.0,                                                                                                                                                          
  "pEpmStr": "kW",                                                                                                                                                      
  "psumCalPec": "1",                                                                                                                                                    
  "insulationResistance": 200.0,                                                                                                                                        
  "dispersionRate": 0.0,                                                                                                                                                
  "sirRealtime": 0,                                                                                                                                                     
  "iLeakLimt": 240,                                                                                                                                                     
  "upvTotal": 0,                                                                                                                                                        
  "upvTotalStr": "V",                                                                                                                                                   
  "ipvTotal": 0,                                                                                                                                                        
  "ipvTotalStr": "A",                                                                                                                                                   
  "powTotal": 0,                                                                                                                                                        
  "powTotalStr": "W",                                                                                                                                                   
  "parallelStatus": 0,                                                                                                                                                  
  "parallelAddr": 0,                                                                                                                                                    
  "parallelPhase": 0,                                                                                                                                                   
  "parallelBattery": 0,                                                                                                                                                 
  "batteryAlarm": "0",                                                                                                                                                  
  "bypassAcOnoffSet": 1.0,                                                                                                                                              
  "bypassAcVoltageSet": 230.0,                                                                                                                                          
  "bypassAcCurrentSet": 50.0,                                                                                                                                           
  "batteryCDEnableSet": 1.0,                                                                                                                                            
  "batteryCDSet": 0.0,                                                                                                                                                  
  "batteryCDISet": 35.800000000000004,                                                                                                                                  
  "batteryCMaxiSet": 70.0,                                                                                                                                              
  "batteryDMaxiSet": 70.0,                                                                                                                                              
  "batteryUvpSet": 40.0,                                                                                                                                                
  "batteryFcvSet": 53.5,                                                                                                                                                
  "batteryAcvSet": 53.5,                                                                                                                                                
  "batteryOvpSet": 59.5,                                                                                                                                                
  "batteryOlvEnableSet": 0.0,                                                                                                                                           
  "batteryLaTemp": -200.1,                                                                                                                                              
  "offGridDDepth": 30.0,                                                                                                                                                
  "epsDDepth": 20.0,                                                                                                                                                    
  "epsSwitchTime": "1000",                                                                                                                                              
  "groupId": ",67541,",                                                                                                                                                 
  "acInType": 0,                                                                                                                                                        
  "energyStorageControl": "1",                                                                                                                                          
  "dsp14Ver": "",                                                                                                                                                       
  "meter1Type": 260,                                                                                                                                                    
  "meter2Type": 0,                                                                                                                                                      
  "meter1SiteHigh": 1,                                                                                                                                                  
  "meter2SiteHigh": 0,                                                                                                                                                  
  "meter1TypeLow": 4,                                                                                                                                                   
  "meter2TypeLow": 0,                                                                                                                                                   
  "generatorPower": 0.0,                                                                                                                                                
  "generatorPowerStr": "kW",                                                                                                                                            
  "generatorPowerPec": "1",                                                                                                                                             
  "generatorTodayEnergy": 0.0,                                                                                                                                          
  "generatorTodayEnergyStr": "kW",                                                                                                                                      
  "generatorTodayEnergyPec": "1",                                                                                                                                       
  "generatorTotalEnergy": 0.0,                                                                                                                                          
  "generatorTotalEnergyStr": "kW",                                                                                                                                      
  "generatorTotalEnergyPec": "1",
  "generatorWarning": "0",                                                                                                                                              
  "generatorSet": "0",                                                                                                                                                  
  "generatorSet01": 0.0,                                                                                                                                                
  "parallelOnoff": "0",                                                                                                                                                 
  "parallelOnoff01": 0.0,                                                                                                                                               
  "parallelOnoff02": 0.0,                                                                                                                                               
  "parallelNumber": 0.0,                                                                                                                                                
  "parallelOnline": 0.0,                                                                                                                                                
  "psumStr": "kW",                                                                                                                                                      
  "psumCal": 0.0,                                                                                                                                                       
  "psumCalStr": "kW",                                                                                                                                                   
  "psum": 0.0,                                                                                                                                                          
  "dcPacStr": "VA",                                                                                                                                                     
  "reactivePowerStr": "Var",                                                                                                                                            
  "apparentPowerStr": "VA",                                                                                                                                             
  "familyLoadPowerPec": "1" 
[/code]

**Comment by ZuinigeRijder** [comment]
Thanks, apparently the documentation of the API is wrong, it should be psum, without capital S.
Unfortunately the grid usage is 0, so I am not sure if below is correct, but you should check when you get power from the grid and when you deliver back to the grid, with above logging. My first guess:

````
"pac": 2.122,
"pacStr": "kW",
= solar (computed to watt in SolisCloud2PVOutput)

"batteryPower": 1.79,
"batteryPowerStr": "kW",
= battery

"psum": 0.0,
"psumStr": "kW",
= fromgrid?

home usage = fromgrid + battery + solar
           = 0        + 2.122   + 1.79
           = 3.912 kW

Note that fromgrid and battery can be negative, I guess.
````

**Reply by cajsoft** [reply]
in the image I sent above, it shows solar generation : 2.121kw and battery pull : 1.571kw.. total 3.692kw

from your explanation above you state it's the opposite way around?

**Comment by ZuinigeRijder** [comment]
Yeah, I did wrongly reverse the values, solar is 2.122 and battery is 1.79.

**Comment by cajsoft** [comment]
think I got it..

**Comment by ZuinigeRijder** [comment]
As stated in my previous post, battery and grid can be negative, so:

home usage = grid + battery + solar
= -1.732 kW + -2.045 kW + 4.152 kW
= 0.375 kW
= 0.375

**Reply by cajsoft** [reply]
thank you.. I Think I misunderstood your post earlier..  I've created the calculation and will monitor the consumption figure throughout the day.

thanks again for all your help... awesome script !

**Comment by ZuinigeRijder** [comment]
I am not sure if the grid value is psum or if this is familyLoadPower.
Let me know, after some days, what you came up with, for the home usage, etc. Including your python code changes.
Maybe I can incorporate this then in the SolisCloud2PVOutput script via a configuration.

**Reply by cajsoft** [reply]
looking good.. had it running for 2 hours and each time the figures match the images.. so all good.

  batteryPower= round(inverter_detail['batteryPower'] * 1000)

        print("Batter Pwr: " + str(batteryPower))

        solarGen = round(inverter_detail['pac'] * 1000)

        print("Solar Gen: " + str(solarGen))

        gridPower = round(inverter_detail['psum'] * 1000)                                        

        print("Grid: " + str(gridPower))

        houseCons = (solarGen - gridPower - batteryPower) / 1000

        print("Load: " + str(houseCons))

**Comment by ZuinigeRijder** [comment]
@cajsoft  I am still puzzled about familyLoadPower. I asked here about the familyLoadPower:

https://github.com/ZuinigeRijder/SolisCloud2PVOutput/issues/15#issuecomment-1555773413

````
Familyload is all power consumed from all sources
````

This does not match with your example, where familyLoadPower is:
https://github.com/ZuinigeRijder/SolisCloud2PVOutput/discussions/24#discussioncomment-5948179

````
"familyLoadPower": 0.179,
"familyLoadPowerStr": "kW",
````

This does not match with your values, at least I do not see it.

**Comment by ZuinigeRijder** [comment]
@cajsoft  [the new version](https://github.com/ZuinigeRijder/SolisCloud2PVOutput/releases/tag/R1.13.0) should now support your use case using configuration. Maybe you can test if this is working for you? Especially because I do not have a solis monitoring solution installation.
